dds/idl/value_reader_generator.cpp

```diff
@@ -135,6 +135,20 @@ namespace {
       indent << "if (!value_reader.end_sequence()) return false;\n";
   }
 
+  void optional_helper(const std::string& expression, AST_Type* type, const std::string& idx, int level)
+  {
+    const std::string indent(level * 2, ' ');
+    be_global->impl_ << indent << "if (!value_reader.begin_optional()) return false;\n";
+
+    be_global->impl_ << indent << "bool has_value = false;\n";
+    be_global->impl_ << indent << "value_reader.read_boolean(has_value);\n";
+    be_global->impl_ << indent << "if (has_value) {\n";
+    generate_read(expression + "().value()", "", type, idx + "i", 4);
+    be_global->impl_ << indent << "}\n";
+
+    be_global->impl_ << indent << "if (!value_reader.end_optional()) return false;\n";
+  }
+
```

Comment:
This is the generated vread function. I opted to wrap the generate_read function rather than modifying it, since it seemed cleaner. However that creates some weirdness with the reads that already create temporary values, like Strings. I could also skip the tmp if it is a string, but I'm not sure if there would be other edge cases, this seemed fine to me

```C++
  while (value_reader.members_remaining()) {
    if (!value_reader.begin_struct_member(member_id, helper)) return false;
    if (!value_reader.member_has_value()) continue;
    switch (member_id) {
    case 0: {
      {
         ::CORBA::Boolean bx;
        if (!value_reader.read_boolean(bx)) return false;
        value.bool_field() = bx;
      }
      break;
    }
    case 1: {
       ::CORBA::Short tmp;
      if (!value_reader.read_int16(tmp)) return false;
        value.short_field() = std::move(tmp);
      break;
    }
    case 2: {
       ::CORBA::Long tmp;
      if (!value_reader.read_int32(tmp)) return false;
        value.int32_field() = std::move(tmp);
      break;
    }
    case 3: {
       ::CORBA::LongLong tmp;
      if (!value_reader.read_int64(tmp)) return false;
        value.int64_field() = std::move(tmp);
      break;
    }
    case 4: {
      String tmp;
      {
        String x;
        if (!value_reader.read_string(x)) return false;
        tmp = x;
      }
        value.str_field() = std::move(tmp);
      break;
    }
    case 5: {
       ::optional::seq_int_t tmp;
      if (!value_reader.begin_sequence(XTypes::TK_INT32)) return false;
      for (size_t ii = 0; value_reader.elements_remaining(); ++ii) {
        tmp.resize(tmp.size() + 1);
        if (!value_reader.begin_element()) return false;
        if (!value_reader.read_int32(tmp[ii])) return false;
        if (!value_reader.end_element()) return false;
      }
      if (!value_reader.end_sequence()) return false;
        value.seq_field() = std::move(tmp);
      break;
    }
    }
    if (!value_reader.end_struct_member()) return false;
  }
  if (!value_reader.end_struct()) return false;
  return true;
}
``

Reply:
It looks like there a redundancy in `optional_helper` compared to `generate_read` that causes the extra temporary variable in the case of string/wstring. I think changing `generate_read` to make it aware of the optionality of the member is okay.